src/modules/keyboardmanager/common/KeyDelay.cpp

```diff
@@ -0,0 +1,165 @@
+#include "pch.h"
+#include "KeyDelay.h"
+
+KeyDelay::~KeyDelay()
+{
+    std::unique_lock<std::mutex> l(_queueMutex);
+    _quit = true;
+    _cv.notify_all();
+    l.unlock();
+    _delayThread.join();
+}
+
+void KeyDelay::KeyEvent(LowlevelKeyboardEvent* ev)
+{
+    std::lock_guard guard(_queueMutex);
+    _queue.push({ ev->lParam->time, ev->wParam });
+    _cv.notify_all();
+}
+
+KeyTimedEvent KeyDelay::NextEvent()
+{
+    auto ev = _queue.front();
+    _queue.pop();
+    return ev;
+}
+
+bool KeyDelay::CheckIfMillisHaveElapsed(DWORD first, DWORD last, DWORD duration)
+{
+    if (first < last && first <= first + duration)
+    {
+        return first + duration < last;
+    }
+    else
+    {
+        first += ULONG_MAX / 2;
+        last += ULONG_MAX / 2;
+        return first + duration < last;
+    }
+}
+
+bool KeyDelay::HasNextEvent()
+{
+    return !_queue.empty();
+}
+
+void KeyDelay::HandleRelease()
+{
+    while (HasNextEvent())
+    {
+        auto ev = NextEvent();
+        switch (ev.message)
+        {
+        case WM_KEYDOWN:
+        case WM_SYSKEYDOWN:
+            _state = KeyDelayState::ON_HOLD;
+            _initialHoldKeyDown = ev.time;
+            return;
+        case WM_KEYUP:
+        case WM_SYSKEYUP:
+            break;
+        }
+    }
+
+    return;
+}
+
+void KeyDelay::HandleOnHold(std::unique_lock<std::mutex>& cvLock)
+{
+    while (HasNextEvent())
+    {
+        auto ev = NextEvent();
+        switch (ev.message)
+        {
+        case WM_KEYDOWN:
+        case WM_SYSKEYDOWN:
+            break;
+        case WM_KEYUP:
+        case WM_SYSKEYUP:
+            if (CheckIfMillisHaveElapsed(_initialHoldKeyDown, ev.time, LONG_PRESS_DELAY_MILLIS))
+            {
+                if (_onLongPressDetected != nullptr)
+                {
+                    _onLongPressDetected(_key);
+                }
+                if (_onLongPressReleased != nullptr)
+                {
+                    _onLongPressReleased(_key);
+                }
+            }
+            else
+            {
+                if (_onShortPress != nullptr)
+                {
+                    _onShortPress(_key);
+                }
+            }
+            _state = KeyDelayState::RELEASED;
+            return;
+        }
+    }
+
+    if (CheckIfMillisHaveElapsed(_initialHoldKeyDown, GetTickCount(), LONG_PRESS_DELAY_MILLIS))

```

Comment:
[GetTickCount](https://docs.microsoft.com/en-us/windows/win32/api/sysinfoapi/nf-sysinfoapi-gettickcount) resets to 0 every 49 days. does that match the behaviour for the `ev.time`?